Library/Homebrew/livecheck/livecheck.rb

```diff
@@ -0,0 +1,384 @@
+        return status_hash(formula, "deprecated", args: args) if args.json?
+
+        unless args.quiet?
+          puts "#{Tty.red}#{formula_name(formula, args: args)}#{Tty.reset} : deprecated"
+          return
+        end
+      end
+
+      if formula.to_s.include?("@") && !formula.livecheckable?
+        return status_hash(formula, "versioned", args: args) if args.json?
+
+        unless args.quiet?
+          puts "#{Tty.red}#{formula_name(formula, args: args)}#{Tty.reset} : versioned"
+          return
+        end
+      end
+
+      if !formula.stable? && !formula.any_version_installed?
+        head_only_msg = "HEAD only formula must be installed to be livecheckable"
+        return status_hash(formula, "error", [head_only_msg], args: args) if args.json?
+
+        unless args.quiet?
+          puts "#{Tty.red}#{formula_name(formula, args: args)}#{Tty.reset} : #{head_only_msg}"
+          return
+        end
+      end
+
+      is_gist = formula.stable&.url&.include?("gist.github.com")
+      if formula.livecheck.skip? || is_gist
+        skip_msg = if formula.livecheck.skip_msg.is_a?(String) &&
+                      !formula.livecheck.skip_msg.blank?
+          formula.livecheck.skip_msg.to_s
+        elsif is_gist
+          "Stable URL is a GitHub Gist"
+        else
+          ""
+        end
+
+        return status_hash(formula, "skipped", (skip_msg.blank? ? nil : [skip_msg]), args: args) if args.json?
+
+        unless args.quiet?
+          puts "#{Tty.red}#{formula_name(formula, args: args)}#{Tty.reset} : skipped" \
+              "#{" - #{skip_msg}" unless skip_msg.blank?}"
+          return
+        end
+      end
+
+      false
+    end
+
+    def print_latest_version(info, args:)
+      formula_s = "#{Tty.blue}#{info["formula"]}#{Tty.reset}"
+      formula_s += " (guessed)" if !info["meta"]["livecheckable"] && args.verbose?
+
+      current_s =
+        if info["version"]["newer_than_upstream"]
+          "#{Tty.red}#{info["version"]["current"]}#{Tty.reset}"
+        else
+          info["version"]["current"]
+        end
+
+      latest_s =
+        if info["version"]["outdated"]
+          "#{Tty.green}#{info["version"]["latest"]}#{Tty.reset}"
+        else
+          info["version"]["latest"]
+        end
+
+      puts "#{formula_s} : #{current_s} ==> #{latest_s}"
+    end
+
+    def checkable_urls(formula)
+      urls = []
+      urls << formula.head.url if formula.head
+      if formula.stable
+        urls << formula.stable.url
+        urls.concat(formula.stable.mirrors)
+      end
+      urls << formula.homepage if formula.homepage
+
+      urls.compact
+    end
+
+    def preprocess_url(url)
+      # Check for GitHub repos on github.com, not AWS
+      url.sub!("github.s3.amazonaws.com", "github.com") if url.include?("github")
+
+      # Use repo from GitHub or GitLab inferred from download URL
+      if url.include?("github.com") && GITHUB_SPECIAL_CASES.none? { |sc| url.include? sc }
```

Comment:
There's some code in `audit.rb` and `utils/notability.rb` that seems similar to this that might be cool to combine?

Reply:
Any thoughts on this?

Reply:
The `GITHUB_SPECIAL_CASES` are simply used to skip the URL preprocessing for specific GitHub URLs (i.e., converting a GitHub URL to the `.git` URL). If a GitHub URL contains one of the strings found in that array, then the URL isn't modified.

I'm not fond of the URL preprocessing as it currently stands (i.e., this should be handled in strategies instead) and I plan to add some GitHub strategies to get rid of this setup not too long after the migrations are finished. These special cases only apply to the context of livecheck and I'm not sure that it would make sense to incorporate them elsewhere, especially considering that they will be removed in the not-too-distant future.

Reply:
Fine with me 👍🏻